In [1]:
import pandas as pd
import sklearn
import numpy as np

a) Laden Sie den Breast Cancer Wisconsin

In [2]:
df=pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/breast-cancer-wisconsin/wdbc.data', header=None)
df = df.drop(columns=0)
df

1      2      3       4       5        6        7        8        9   \
0    M  17.99  10.38  122.80  1001.0  0.11840  0.27760  0.30010  0.14710   
1    M  20.57  17.77  132.90  1326.0  0.08474  0.07864  0.08690  0.07017   
2    M  19.69  21.25  130.00  1203.0  0.10960  0.15990  0.19740  0.12790   
3    M  11.42  20.38   77.58   386.1  0.14250  0.28390  0.24140  0.10520   
4    M  20.29  14.34  135.10  1297.0  0.10030  0.13280  0.19800  0.10430   
..  ..    ...    ...     ...     ...      ...      ...      ...      ...   
564  M  21.56  22.39  142.00  1479.0  0.11100  0.11590  0.24390  0.13890   
565  M  20.13  28.25  131.20  1261.0  0.09780  0.10340  0.14400  0.09791   
566  M  16.60  28.08  108.30   858.1  0.08455  0.10230  0.09251  0.05302   
567  M  20.60  29.33  140.10  1265.0  0.11780  0.27700  0.35140  0.15200   
568  B   7.76  24.54   47.92   181.0  0.05263  0.04362  0.00000  0.00000   

         10  ...      22     23      24      25       26       27      28  \
0    0.2419  ...  25.380  17.33  184.60  2019.0  0.16220  0.66560  0.7119   
1    0.1812  ...  24.990  23.41  158.80  1956.0  0.12380  0.18660  0.2416   
2    0.2069  ...  23.570  25.53  152.50  1709.0  0.14440  0.42450  0.4504   
3    0.2597  ...  14.910  26.50   98.87   567.7  0.20980  0.86630  0.6869   
4    0.1809  ...  22.540  16.67  152.20  1575.0  0.13740  0.20500  0.4000   
..      ...  ...     ...    ...     ...     ...      ...      ...     ...   
564  0.1726  ...  25.450  26.40  166.10  2027.0  0.14100  0.21130  0.4107   
565  0.1752  ...  23.690  38.25  155.00  1731.0  0.11660  0.19220  0.3215   
566  0.1590  ...  18.980  34.12  126.70  1124.0  0.11390  0.30940  0.3403   
567  0.2397  ...  25.740  39.42  184.60  1821.0  0.16500  0.86810  0.9387   
568  0.1587  ...   9.456  30.37   59.16   268.6  0.08996  0.06444  0.0000   

         29      30       31  
0    0.2654  0.4601  0.11890  
1    0.1860  0.2750  0.08902  
2    0.2430  0.3613  0.08758  
3    0.2575  0.6638  0.17300  
4    0.1625  0.2364  0.07678  
..      ...     ...      ...  
564  0.2216  0.2060  0.07115  
565  0.1628  0.2572  0.06637  
566  0.1418  0.2218  0.07820  
567  0.2650  0.4087  0.12400  
568  0.0000  0.2871  0.07039  

[569 rows x 31 columns]

Insgesamt gibt es **31 Feature** und **1 Target** in dem Datensatz. Das Target ist dabei entweder **M (malignant)** oder **B (benign)**.</br>
Im nächsten Schritt werden diese Label in numerische Attribute umgewandelt.


b) Nutzen Sie die sklearn.preprocessing.LabelEncoder um die binary Label in ein numerisches
Attribut zu kodieren.


Nach dem Kodieren ist:

**M = 1** </br>
**B = 0** </br>

In [3]:
#Use LabelEncoder to transform the Lable into numeric features
from sklearn.preprocessing import LabelEncoder
lb_make = LabelEncoder()
df[1] = lb_make.fit_transform(df[1])
df

1      2      3       4       5        6        7        8        9   \
0     1  17.99  10.38  122.80  1001.0  0.11840  0.27760  0.30010  0.14710   
1     1  20.57  17.77  132.90  1326.0  0.08474  0.07864  0.08690  0.07017   
2     1  19.69  21.25  130.00  1203.0  0.10960  0.15990  0.19740  0.12790   
3     1  11.42  20.38   77.58   386.1  0.14250  0.28390  0.24140  0.10520   
4     1  20.29  14.34  135.10  1297.0  0.10030  0.13280  0.19800  0.10430   
..   ..    ...    ...     ...     ...      ...      ...      ...      ...   
564   1  21.56  22.39  142.00  1479.0  0.11100  0.11590  0.24390  0.13890   
565   1  20.13  28.25  131.20  1261.0  0.09780  0.10340  0.14400  0.09791   
566   1  16.60  28.08  108.30   858.1  0.08455  0.10230  0.09251  0.05302   
567   1  20.60  29.33  140.10  1265.0  0.11780  0.27700  0.35140  0.15200   
568   0   7.76  24.54   47.92   181.0  0.05263  0.04362  0.00000  0.00000   

         10  ...      22     23      24      25       26       27      28  \
0    0.2419  ...  25.380  17.33  184.60  2019.0  0.16220  0.66560  0.7119   
1    0.1812  ...  24.990  23.41  158.80  1956.0  0.12380  0.18660  0.2416   
2    0.2069  ...  23.570  25.53  152.50  1709.0  0.14440  0.42450  0.4504   
3    0.2597  ...  14.910  26.50   98.87   567.7  0.20980  0.86630  0.6869   
4    0.1809  ...  22.540  16.67  152.20  1575.0  0.13740  0.20500  0.4000   
..      ...  ...     ...    ...     ...     ...      ...      ...     ...   
564  0.1726  ...  25.450  26.40  166.10  2027.0  0.14100  0.21130  0.4107   
565  0.1752  ...  23.690  38.25  155.00  1731.0  0.11660  0.19220  0.3215   
566  0.1590  ...  18.980  34.12  126.70  1124.0  0.11390  0.30940  0.3403   
567  0.2397  ...  25.740  39.42  184.60  1821.0  0.16500  0.86810  0.9387   
568  0.1587  ...   9.456  30.37   59.16   268.6  0.08996  0.06444  0.0000   

         29      30       31  
0    0.2654  0.4601  0.11890  
1    0.1860  0.2750  0.08902  
2    0.2430  0.3613  0.08758  
3    0.2575  0.6638  0.17300  
4    0.1625  0.2364  0.07678  
..      ...     ...      ...  
564  0.2216  0.2060  0.07115  
565  0.1628  0.2572  0.06637  
566  0.1418  0.2218  0.07820  
567  0.2650  0.4087  0.12400  
568  0.0000  0.2871  0.07039  

[569 rows x 31 columns]

c) Mit sklearn.cross_validation.train_test_split splitten Sie die Daten in Trainings (80%) und
Testdaten (20%). Setzen Sie random_state=1.

In [4]:
#Split the data into 80% train and 20% test data 
from sklearn.model_selection import train_test_split
data = df.drop(columns=1)
target= df[1]
x_train, x_test, y_train, y_test = train_test_split(data, target, test_size=0.2, random_state=1)

d) Als Preprocessingschritte nutzen Sie sklearn.preprocessing.StandardScaler für feature scaling
und für Dimensionalitätsreduktion ein PCA mit sklearn.decomposition. PCA und mit
n_components=2. Als Klassifikator nutzen Sie sklearn.linear_model.LogisticRegression mit
random_state=1. Packen Sie alle Schritte in eine Pipeline (sklearn.pipeline.Pipeline).

In [5]:
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.decomposition import PCA
from sklearn.linear_model import LogisticRegression

In [6]:
#Setup a pipeline, which standardize the data for a pca and applies a logistic regression on the data
pipe = Pipeline(steps=[
    ('scaler', StandardScaler()),('pca',PCA(n_components=2)),('clf',LogisticRegression(random_state=1))])
pipe.fit(x_train,y_train)

Pipeline(steps=[('scaler', StandardScaler()), ('pca', PCA(n_components=2)),
                ('clf', LogisticRegression(random_state=1))])

e) Testen Sie mittels pipeline.score die Genauigkeit des Modells. Sie sollten eine Genauigkeit
von 0. 947 erreichen.

In [7]:
pipe.score(x_test,y_test)

0.9473684210526315

Die Genauigkeit der Pipeline beträgt **94%** (0.947)


f) Statt PCA nutzen Sie ein Recursive Feature Elimination (RFE) zur Feature-Auswahl
(sklearn.feature_selection.RFECV). Wie viele und welche Features sind für die Klassifikation
interessant? Welche (max.) Genauigkeit auf die Testdaten kann mit diesem Schritt (statt PCA)
erreicht werden?

In [8]:
#Use the pipeline with a rfecv instead of a pca and show wich features are important.
from sklearn.feature_selection import RFECV
pipe2 = Pipeline(steps=[
    ('scaler', StandardScaler()),('rfecv',RFECV(estimator=LogisticRegression(random_state=1)))])
pipe2.fit(x_train,y_train)
pipe2['rfecv'].support_

array([ True,  True,  True,  True, False,  True,  True,  True, False,
       False,  True, False,  True,  True, False,  True, False,  True,
       False,  True,  True,  True,  True,  True,  True, False,  True,
        True,  True,  True])

In [9]:
print('Anzahl der interessanten Feature:',sum(pipe2['rfecv'].support_==True))

Anzahl der interessanten Feature: 22


In [10]:
imp_feature=x_test.columns.where(pipe2['rfecv'].support_==True).dropna().tolist()
print('Die Indizes der wichtigen Feature sind:',imp_feature)

Die Indizes der wichtigen Feature sind: [2.0, 3.0, 4.0, 5.0, 7.0, 8.0, 9.0, 12.0, 14.0, 15.0, 17.0, 19.0, 21.0, 22.0, 23.0, 24.0, 25.0, 26.0, 28.0, 29.0, 30.0, 31.0]


In [11]:
pipe2.score(x_test,y_test)

0.9824561403508771

Insgesamt werden beim RFE 22 von 31 Featuren benutzt.

Dabei erreicht die Pipeline eine Genauigkeit von **98%**(0.982), was ca. **4%** besser ist als mit der PCA-Methode.